lib/semantics/assignment.cc

```diff
@@ -563,6 +558,117 @@ void AssignmentContext::CheckForImpureCall(
   }
 }
 
+// C1594 checks
+static bool IsPointerDummyOfPureFunction(const Symbol &x) {
+  return IsPointerDummy(x) && FindPureProcedureContaining(x.owner()) &&
+      x.owner().symbol() && IsFunction(*x.owner().symbol());
+}
+
+static const char *WhyBaseObjectIsSuspicious(
+    const Symbol &x, const Scope &scope) {
+  // See C1594, first paragraph.  These conditions enable checks on both
+  // left-hand and right-hand sides in various circumstances.
+  if (IsHostAssociated(x, scope)) {
+    return "host-associated";
+  } else if (IsUseAssociated(x, scope)) {
+    return "USE-associated";
+  } else if (IsPointerDummyOfPureFunction(x)) {
+    return "a POINTER dummy argument of a PURE function";
+  } else if (IsIntentIn(x)) {
+    return "an INTENT(IN) dummy argument";
+  } else if (FindCommonBlockContaining(x)) {
+    return "in a COMMON block";
+  } else {
+    return nullptr;
+  }
+}
+
+// Checks C1594(1,2)
+void CheckDefinabilityInPureScope(parser::ContextualMessages &messages,
+    const Symbol &lhs, const Scope &scope) {
+  if (const char *why{WhyBaseObjectIsSuspicious(lhs, scope)}) {
+    evaluate::SayWithDeclaration(messages, &lhs,
+        "A PURE subprogram may not define '%s' because it is %s"_err_en_US,
+        lhs.name(), why);
+  }
+}
+
+static std::optional<std::string> GetPointerComponentDesignatorName(
+    const evaluate::Expr<evaluate::SomeType> &expr) {
+  if (auto type{evaluate::DynamicType::From(expr)}) {
+    if (type->category() == TypeCategory::Derived &&
+        !type->IsUnlimitedPolymorphic()) {
+      UltimateComponentIterator ultimates{type->GetDerivedTypeSpec()};
+      if (auto pointer{
+              std::find_if(ultimates.begin(), ultimates.end(), IsPointer)}) {
+        return pointer.BuildResultDesignatorName();
+      }
+    }
+  }
+  return std::nullopt;
+}
+
+// Checks C1594(5,6)
+void CheckCopyabilityInPureScope(parser::ContextualMessages &messages,
+    const evaluate::Expr<evaluate::SomeType> &expr, const Scope &scope) {
+  if (const Symbol * base{GetFirstSymbol(expr)}) {
+    if (const char *why{WhyBaseObjectIsSuspicious(*base, scope)}) {
+      if (auto pointer{GetPointerComponentDesignatorName(expr)}) {
+        evaluate::SayWithDeclaration(messages, base,
+            "A PURE subprogram may not copy the value of '%s' because it is %s and has the POINTER component '%s'"_err_en_US,
+            base->name(), why, *pointer);
+      }
+    }
+  }
+}
+
+void AssignmentContext::CheckForPureContext(
+    const evaluate::Expr<evaluate::SomeType> &lhs,
+    const evaluate::Expr<evaluate::SomeType> &rhs, parser::CharBlock source,
```

Comment:
"lhs" and "rhs" could be defined as "&SomeExpr".

Reply:
Yes.  Do you want me to do so?

Reply:
`SomeExpr &`; done.